### PR TITLE
[WIP] - Develop psr4 support

### PR DESCRIPTION
**Pull request by steverhoades**
I'm putting it here for review and refinement.  The idea is that you add the following to your module.config.php file (or in an autoload/apigility.local.php file).
```
return [
    'zf-apigility-admin' => [
        'path_spec' => 'psr-4'
    ]
];
```
The default is PSR-0.

### CODE REVIEW

**Comment by weierophinney** [comment]
I love this idea, and the implementation makes a ton of sense.

Now, here's where things go awry:

- First, I need a rebase.
- Second, there's another consideration: zf-apigility-doctrine. I'm not sure if any of its admin model classes inherit from any of these, but if they do, there will be a breakage in the signatures. This will need to be addressed if that's the case.

Any chance you can get the first done, and review the second by late this week?

**Comment by steverhoades** [comment]
@weierophinney I rebased but It looks like I did something wrong.  When I get a chance I'll clean it up.

We aren't using any of the doctrine features, I'll attempt to get an environment up this week to test this out.

**Comment by steverhoades** [comment]
@weierophinney zf-apigility-doctrine would indeed need to be updated as the PSR-0 paths are harded coded (similar to the existing apigility-admin).

**Comment by weierophinney** [comment]
@steverhoades Ping me if you need help with the rebase. :)

**Comment by steverhoades** [comment]
@weierophinney i seem to be making it worse LOL. 2nd attempt failed.

File: src/Model/ModulePathSpec.php [hunk 1]
```diff
@@ -0,0 +1,177 @@
+<?php
+
+namespace ZF\Apigility\Admin\Model;
+
+use \InvalidArgumentException;
```
Comment (weierophinney) [inline src/Model/ModulePathSpec.php]: No need to globally qualify in import statements.

File: src/Model/ModulePathSpec.php [hunk 2]
```diff
@@ -0,0 +1,177 @@
+<?php
+
```
Comment (weierophinney) [inline src/Model/ModulePathSpec.php]: Can you add the license docblock, please?

File: src/Model/ModulePathSpec.php [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+namespace ZF\Apigility\Admin\Model;
+
+use \InvalidArgumentException;
+use ZF\Configuration\ModuleUtils;
+
+class ModulePathSpec
```
Comment (weierophinney) [inline src/Model/ModulePathSpec.php]: Please add a class-level docblock with a brief summary of the class purpose.

File: src/Model/ModulePathSpec.php [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+{
+    protected $modules;
```
Comment (weierophinney) [inline src/Model/ModulePathSpec.php]: Please document all properties.

**Comment by weierophinney** [comment]
I did a quick look over zfcampus/zf-apigility-doctrine, and it looks like it's using `ModuleUtils`. As such, it should continue to work, but it won't be able to generate or use PSR-4 skeletons until it's updated to use this code. If you can work on that, the path would be:

- Clone zfcampus/zf-apigility-doctrine
- Create a new branch for the feature
- Modify the `composer.json` to:
  - Add a `repositories` section
  - which adds a "vcs" repository pointing to your own repository for zf-apigility-admin
  - and modify the `zfcampus/zf-apigility-admin` dependency to use `dev-develop-psr4@dev` as the version

Make the updates, and then once this is merged, you can point it to the develop branch locally.

Also, per our conversation on IRC, the `ModulePathSpec` class should have some unit tests.

Great work, @steverhoades !

**Comment by weierophinney** [comment]
@steverhoades Looks great! I'll merge today; can you work on the zf-apigility-doctrine changes, too? It looks like we'll be releasing early next week.

**Comment by steverhoades** [comment]
@weierophinney I have started on the doctrine changes.  Hope to have something to review later today.

**Comment by weierophinney** [comment]
@steverhoades Awesome, thanks!

**Comment by weierophinney** [comment]
Merged to develop for release with 1.1.

**Pull request merged**